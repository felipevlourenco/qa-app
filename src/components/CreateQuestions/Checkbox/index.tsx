import React from 'react'
import { CheckboxProp } from 'components/CreateQuestions/Checkbox/types'
import styles from 'components/CreateQuestions/Checkbox/styles.module.css'

const Checkbox = ({ checked, onChange }: CheckboxProp) => {
  return (
    <label className={styles.label} htmlFor="delay">
      <input id="delay" type="checkbox" checked={checked} onChange={onChange} className={styles.checkbox} />
      <span className={styles.span}>Add question with delay</span>
    </label>
  )
}

export default Checkbox
