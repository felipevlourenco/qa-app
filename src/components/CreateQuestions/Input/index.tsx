import React from 'react'
import { InputProp } from 'components/CreateQuestions/Input/types'
import styles from 'components/CreateQuestions/Input/styles.module.css'

const Input = ({ id, label, errors, register }: InputProp) => {
  return (
    <div className={styles.field}>
      <label className={styles.label} htmlFor={id}>
        {label}
      </label>
      <input id={id} {...register(id, { required: true })} className={styles.input} />
      {errors[id] && <span className={styles.error}>This field is required</span>}
    </div>
  )
}

export default Input
