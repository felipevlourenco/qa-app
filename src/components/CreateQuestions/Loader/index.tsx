import { useEffect, useState } from 'react'
import ReactDOM from 'react-dom'
import styles from 'components/CreateQuestions/Loader/styles.module.css'

export const Loader = ({ className = 'root-portal', element = 'div' }) => {
  const [container] = useState(() => {
    const el = document.createElement(element)
    el.classList.add(className)
    return el
  })

  useEffect(() => {
    document.body.appendChild(container)
    return () => {
      document.body.removeChild(container)
    }
  }, [])

  return ReactDOM.createPortal(
    <div className={styles.wrapper}>
      <div className={styles.loader} />
      <span className={styles.text}>Adding question...</span>
    </div>,
    container
  )
}

export default Loader
