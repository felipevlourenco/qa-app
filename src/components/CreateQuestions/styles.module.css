.create {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  margin-top: 1rem;
}

.header {
  font-size: 1.2rem;
  font-weight: 600;
}

.form {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  margin-top: 1rem;
}