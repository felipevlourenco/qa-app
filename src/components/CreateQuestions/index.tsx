import React from 'react'
import { useForm } from 'react-hook-form'
import styles from 'components/CreateQuestions/styles.module.css'
import Button from 'components/Button'
import { AddQuestionAction } from 'store/questions/types'
import { useAppDispatch, useAppSelector } from 'hooks/useRedux'
import { addQuestion, addQuestionAsyncAction } from 'store/questions'
import Input from 'components/CreateQuestions/Input'
import Checkbox from 'components/CreateQuestions/Checkbox'
import Loader from 'components/CreateQuestions/Loader'
import useToggle from 'hooks/useToggle'
import Tooltip from 'components/Tooltip'

const CreateQuestions = () => {
  const dispatch = useAppDispatch()
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors }
  } = useForm<AddQuestionAction>()
  const { toggle, isToggled } = useToggle()
  const { state } = useAppSelector((questionState) => questionState.questions)

  const onSubmit = handleSubmit((data) => {
    dispatch(isToggled ? addQuestionAsyncAction(data) : addQuestion(data))
    setValue('question', '')
    setValue('answer', '')
  })

  return (
    <div className={styles.create}>
      <Tooltip content="Here you can create new questions and their answers">
        <span className={styles.header}>Create a new question</span>
      </Tooltip>
      <form className={styles.form} onSubmit={onSubmit}>
        <Input id="question" label="Question" register={register} errors={errors} />
        <Input id="answer" label="answer" register={register} errors={errors} />
        <Checkbox checked={isToggled} onChange={toggle} />
        <Button type="submit">Create question</Button>
      </form>

      {state === 'PENDING' && <Loader />}
    </div>
  )
}

export default CreateQuestions
