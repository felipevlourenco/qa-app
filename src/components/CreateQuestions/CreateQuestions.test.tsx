import { fireEvent, render, queryByAttribute, act, waitFor } from '@testing-library/react'
import CreateQuestions from 'components/CreateQuestions'
import MockStore from 'test/utils/MockStore'

describe('component/CreateQuestions', () => {
  const getById = queryByAttribute.bind(null, 'id')

  it('should render', () => {
    const { container } = render(
      <MockStore>
        <CreateQuestions />
      </MockStore>
    )
    expect(container).toMatchSnapshot()
  })

  it('should add a question', async () => {
    const { container, getByRole } = render(
      <MockStore>
        <CreateQuestions />
      </MockStore>
    )

    const questionInput = getById(container, 'question')
    const answerInput = getById(container, 'answer')
    const button = getByRole('button')

    fireEvent.input(questionInput!, 'question 1')
    fireEvent.input(answerInput!, 'answer 1')

    act(() => {
      fireEvent.click(button)
    })

    await waitFor(() => {
      expect(questionInput).toHaveValue('')
      expect(answerInput).toHaveValue('')
    })
  })
})
