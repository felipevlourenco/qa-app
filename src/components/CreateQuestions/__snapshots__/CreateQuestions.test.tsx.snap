// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/CreateQuestions should render 1`] = `
<div>
  <div
    class="create"
  >
    <div
      class="tooltip"
    >
      <div
        class="child"
      >
        <span
          class="header"
        >
          Create a new question
        </span>
      </div>
    </div>
    <form
      class="form"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="question"
        >
          Question
        </label>
        <input
          class="input"
          id="question"
          name="question"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="answer"
        >
          answer
        </label>
        <input
          class="input"
          id="answer"
          name="answer"
          value=""
        />
      </div>
      <label
        class="label"
        for="delay"
      >
        <input
          class="checkbox"
          id="delay"
          type="checkbox"
        />
        <span
          class="span"
        >
          Add question with delay
        </span>
      </label>
      <button
        class="button primary"
        type="submit"
      >
        Create question
      </button>
    </form>
  </div>
</div>
`;
